src/sql/src/func.rs

```diff
@@ -906,10 +949,154 @@ impl From<ScalarBaseType> for ParamType {
+    fn return_type(&self, ecx: &ExprContext, param_list: &ParamList) -> ReturnType {
+        fn assert_oti_len(oti: &[ColumnType], len: usize, name: &str) {
+            assert_eq!(
+                oti.len(),
+                len,
+                "{} requires exactly {} contrived input to automatically determine return type",
+                name,
+                len,
+            );
+        }
+
+        let mut output_type_inputs: Vec<ColumnType> = param_list
+            .contrive_coercible_exprs()
+            .into_iter()
+            .map(|c| {
+                let expr = c.type_as_any(&ecx).expect("c is typed NULL");
+                ecx.column_type(&expr)
+            })
+            .collect();
+
+        let c = match self {
+            HirScalarExpr::Literal(_row, column_type) => column_type.clone(),
+            HirScalarExpr::CallNullary(func) => {
+                assert_oti_len(&output_type_inputs, 0, "HirScalarExpr::CallNullary");
+                func.output_type()
             }
-        })
+            HirScalarExpr::CallUnary { func, .. } => {
+                assert_oti_len(&output_type_inputs, 1, "HirScalarExpr::CallUnary");
+                func.output_type(output_type_inputs.remove(0))
+            }
+            HirScalarExpr::CallBinary { func, .. } => {
+                assert_oti_len(&output_type_inputs, 2, "HirScalarExpr::CallBinary");
+                func.output_type(output_type_inputs.remove(0), output_type_inputs.remove(0))
+            }
+            HirScalarExpr::CallVariadic { func, .. } => func.output_type(output_type_inputs),
+            other => unreachable!(
+                "unexepected HirScalarExpr in Operation<HirScalarExpr>::return_type: {:?}",
+                other
+            ),
+        };
+
+        ReturnType::scalar(c.scalar_type.into())
+    }
+}
+
+impl GetReturnType for (HirScalarExpr, AggregateFunc) {
+    fn return_type(&self, ecx: &ExprContext, _param_list: &ParamList) -> ReturnType {
+        let c = ecx.column_type(&self.0);
+        let s = self.1.output_type(c).scalar_type;
+        ReturnType::scalar(s.into())
+    }
+}
+
+impl GetReturnType for ScalarWindowFunc {
+    fn return_type(&self, _ecx: &ExprContext, _param_list: &ParamList) -> ReturnType {
+        ReturnType::scalar(self.output_type().scalar_type.into())
+    }
+}
+
+impl GetReturnType for TableFuncPlan {
+    fn return_type(&self, _ecx: &ExprContext, _param_list: &ParamList) -> ReturnType {
+        let mut cols: Vec<ScalarType> = match &self.expr {
+            HirRelationExpr::CallTable { func, .. } => func
+                .output_type()
+                .column_types
+                .into_iter()
+                .map(|col| col.scalar_type)
+                .collect(),
+            other => unreachable!(
+                "unexepected HirRelationExpr in Operation<TableFuncPlan>::return_type: {:?}",
+                other
+            ),
+        };
+
+        match cols.len() {
```

Comment:
Ah, clever! This means we require a clone because we don't have ownership of the borrowed items and they're `Clone` not `Copy`, though, so trading potential ownership issues w/ unnecessary clones. Because this is the only requested code change, going to go with the version as is to avoid going back through CI--glad to change, though, if you feel strongly.